src/ripple/overlay/Slot.h

```diff
@@ -0,0 +1,635 @@
+     * @param peerPtr Peer which received the message
+     * @param type  Message type (Validation and Propose Set only,
+     *     others are ignored, future use)
+     */
+    void
+    update(
+        PublicKey const& validator,
+        id_t id,
+        std::weak_ptr<Peer> const& peerPtr,
+        protocol::MessageType type);
+
+    /** Handle peer deletion when a peer disconnects.
+     * If the peer is in Selected state then
+     * call unsquelch handler for every peer in squelched state and reset
+     * every peer's state to Counting. Switch Slot's state to Counting.
+     * @param validator Public key of the source validator
+     * @param id Deleted peer id
+     * @param erase If true then erase the peer. The peer is not erased
+     *      when the peer when is idled. The peer is deleted when it
+     *      disconnects
+     */
+    void
+    deletePeer(PublicKey const& validator, id_t id, bool erase);
+
+    /** Get the time of the last peer selection round */
+    const time_point&
+    getLastSelected() const
+    {
+        return lastSelected_;
+    }
+
+    /** Return number of peers in state */
+    std::uint16_t
+    inState(PeerState state) const;
+
+    /** Return number of peers not in state */
+    std::uint16_t
+    notInState(PeerState state) const;
+
+    /** Return Slot's state */
+    SlotState
+    getState() const
+    {
+        return state_;
+    }
+
+    /** Return selected peers */
+    std::set<id_t>
+    getSelected() const;
+
+    /** Get peers info. Return map of peer's state, count, squelch
+     * expiration milsec, and last message time milsec.
+     */
+    std::
+        unordered_map<id_t, std::tuple<PeerState, uint16_t, uint32_t, uint32_t>>
+        getPeers() const;
+
+    /** Check if peers stopped relaying messages. If a peer is
+     * selected peer then call unsquelch handler for all
+     * currently squelched peers and switch the slot to
+     * Counting state.
+     * @param validator Public key of the source validator
+     */
+    void
+    checkIdle(PublicKey const& validator);
+
+private:
+    /** Reset counts of peers in Selected or Counting state */
+    void
+    resetCounts();
+
+    /** Initialize slot to Counting state */
+    void
+    initCounting();
+
+    /** Data maintained for each peer */
+    struct PeerInfo
+    {
+        std::weak_ptr<Peer> const peer;  // peer's weak_ptr
+        PeerState state;                 // peer's state
+        std::size_t count;               // message count
+        time_point expire;               // squelch expiration time
+        time_point lastMessage;          // time last message received
+    };
+    std::unordered_map<id_t, PeerInfo> peers_;  // peer's data
+    std::unordered_set<id_t> considered_;       // pool of peers considered
+                                                // as the source of messages
+                                                // from validator - peers
```

Comment:
Putting very short comments to the right is OK, but when the comment on the right spans several lines I'd rather put the comment on it's own line above the code it's commenting, where it would only take one line instead of three. If you prefer this style then no change needed. Mine is only a preference.